sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +180,67 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
     return new ShardRecordsIterator(checkpoint, kinesis);
   }
 
+  /**
+   * Waits until all records read from given shardRecordsIterator are taken from
+   * {@link #recordsQueue} and acked.
+   * Uses {@link #counters} map to track the amount of remaining events.
+   */
+  private void waitUntilAllShardRecordsRead(ShardRecordsIterator shardRecordsIterator)
+      throws InterruptedException {
+    while (!allShardRecordsRead(shardRecordsIterator)) {
+      Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+    }
+  }
+
+  private boolean allShardRecordsRead(final ShardRecordsIterator shardRecordsIterator) {
+    return counters.get(shardRecordsIterator.getShardId()).get() == 0;
+  }
+
+  private void readFromSuccessiveShards(final ShardRecordsIterator closedShardIterator)
+      throws TransientKinesisException {
+    List<ShardRecordsIterator> successiveShardRecordIterators = closedShardIterator
+        .findSuccessiveShardRecordIterators();
+
+    if (!successiveShardRecordIterators.isEmpty()) {
+
+      // Handle one closedShardIterator at a time to keep shardIteratorsMap,
+      // recordsQueue and counters in sync.
+      synchronized (this) {
+        ImmutableMap.Builder<String, ShardRecordsIterator> shardsMap = ImmutableMap.builder();
+        Iterable<ShardRecordsIterator> allShards = Iterables
+            .concat(shardIteratorsMap.values(), successiveShardRecordIterators);
+        for (ShardRecordsIterator iterator : allShards) {
```

Comment:
What is this loop doing? Simply creating new map with closed shard removed? Can't you simply do:
`shardsMap.remove(closedShardIterator.getShardId())`?

Reply:
Yes, it does exactly that. I could simply remove the shards from the map but it would need to be a thread-safe modification (so ConcurrentMap could be used). Using immutable map seemed a more straightforward solution.

Reply:
Why not use a regular map .remove() but under a synchronized block? You already have a synchronized block here.

Reply:
I don’t think it’s safe to use a regular map here. It’s the only place where synchronized block is used and not all threads will synchronize here. For example the Beam’s thread trying to create a new checkpoint will need to see the up-to-date values in this map.